vmware_guest: fix dvs idempotency issue

reversecipher opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Fixes https://github.com/ansible/ansible/issues/34105. DVS port connection should only be changed if nic change is detected.
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
vmware_guest
##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.5.0 (devel 2e76c89910) last updated 2017/12/20 19:28:48 (GMT +300)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /root/ansible-dev/lib/ansible
  executable location = /root/ansible-dev/bin/ansible
  python version = 2.7.5 (default, Aug  4 2017, 00:39:18) [GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```
```

reversecipher committed:
vmware_guest: fix dvs idempotency issue

Akasurde commented:
@reversecipher Thanks for this PR. Could you please add integration test for this change ?

Akasurde commented:
cc @tchernomax @angystardust @aperigault

tchernomax commented:
@Akasurde @reversecipher I think it would be clearer to re-write this entire section like that:
```
# VDS switch
pg_obj = find_obj(self.content, [vim.dvs.DistributedVirtualPortgroup], network_devices[key]['name'])

if (nic.device.backing and not hasattr(nic.device.backing, 'port')) or \
   (nic.device.backing and (nic.device.backing.port.portgroupKey != pg_obj.key or
                              nic.device.backing.port.switchUuid != pg_obj.config.distributedVirtualSwitch.uuid)):
    dvs_port_connection = vim.dvs.PortConnection()
    dvs_port_connection.portgroupKey = pg_obj.key
    dvs_port_connection.switchUuid = pg_obj.config.distributedVirtualSwitch.uuid
    nic.device.backing = vim.vm.device.VirtualEthernetCard.DistributedVirtualPortBackingInfo()
    nic.device.backing.port = dvs_port_connection
    nic_change_detected = True
```

What do you think ?

Else I am OK with the content of this PR.

On the integration test, have a look at:
* https://github.com/tchernomax/ansible/tree/devel/test/integration/targets/vmware_guest/tasks
* https://github.com/ansible/ansible/pull/30582#issuecomment-330992482

reversecipher commented:
@tchernomax Yes I think that would be clearer. I can update the PR if you like.

I tried creating integration test for the change but run into a issue. Here is the test I tried to run:

```yaml
- name: create new VMs
  vmware_guest:
    validate_certs: False
    hostname: "{{ vcsim }}"
    username: "{{ vcsim_instance['json']['username'] }}"
    password: "{{ vcsim_instance['json']['password'] }}"
    name: "{{ 'newvm_' + item|basename }}"
    guest_id: centos64Guest
    datacenter: "{{ (item|basename).split('_')[0] }}"
    hardware:
        num_cpus: 4
        num_cpu_cores_per_socket: 2
        memory_mb: 512
    disk:
        - size: 0gb
          type: thin
          autoselect_datastore: True
    networks:
        - name: "DC0_DVPG0"
    state: present
    folder: "{{ item|dirname }}"
  with_items: "{{ vmlist['json'] }}"
  register: clone_d1_c1_f0

- debug: var=clone_d1_c1_f0

- name: assert that changes were made
  assert:
    that:
        - "clone_d1_c1_f0.results|map(attribute='changed')|unique|list == [true]"

- name: create new VMs again
  vmware_guest:
    validate_certs: False
    hostname: "{{ vcsim }}"
    username: "{{ vcsim_instance['json']['username'] }}"
    password: "{{ vcsim_instance['json']['password'] }}"
    name: "{{ 'newvm_' + item|basename }}"
    guest_id: centos64Guest
    datacenter: "{{ (item|basename).split('_')[0] }}"
    hardware:
        num_cpus: 4
        num_cpu_cores_per_socket: 2
        memory_mb: 512
    disk:
        - size: 0gb
          type: thin
          autoselect_datastore: True
    networks:
        - name: "DC0_DVPG0"
    state: present
    folder: "{{ item|dirname }}"
  with_items: "{{ vmlist['json'] }}"
  register: clone_d1_c1_f0_recreate

- debug: var=clone_d1_c1_f0_recreate

- name: assert that no changes were made after re-creating
  assert:
    that:
        - "clone_d1_c1_f0_recreate.results|map(attribute='changed')|unique|list == [false]"
```
The creation of vms succeeds, but the re-creation fails with "*types.InvalidDeviceSpec" error:
```
TASK [vmware_guest : create new VMs again] *************************************
failed: [testhost] (item=/DC0/vm/DC0_H0_VM0) => {"changed": true, "item": "/DC0/vm/DC0_H0_VM0", "msg": "*types.InvalidDeviceSpec"}
failed: [testhost] (item=/DC0/vm/DC0_H0_VM1) => {"changed": true, "item": "/DC0/vm/DC0_H0_VM1", "msg": "*types.InvalidDeviceSpec"}
failed: [testhost] (item=/DC0/vm/DC0_C0_RP0_VM0) => {"changed": true, "item": "/DC0/vm/DC0_C0_RP0_VM0", "msg": "*types.InvalidDeviceSpec"}
failed: [testhost] (item=/DC0/vm/DC0_C0_RP0_VM1) => {"changed": true, "item": "/DC0/vm/DC0_C0_RP0_VM1", "msg": "*types.InvalidDeviceSpec"}
```
These tasks succeeded when I tested against real Vcenter. Could this be issue with vcsim?

Akasurde commented:
@reversecipher Yes, you are correct. This can be issue with vcsim.

reversecipher commented:
I tested with both centos7 and ubuntu1604 containers. Shouldn't the vcenter-simulator container get the latest vcsim version?

tchernomax commented:
@reversecipher 

> I can update the PR if you like.

Yes please.

> The creation of vms succeeds, but the re-creation fails with "*types.InvalidDeviceSpec" error…

It seems you didn't copy all the create_d1_c1_f0.yml file, that's why you get an error.
Simpler: modify the create_d1_c1_f0.yml file.
Either way, it seems we will not be able to test this PR with vcsim.
When I print:
```
            dvps = self.cache.get_all_objs(self.content, [vim.dvs.DistributedVirtualPortgroup])
            for dvp in dvps:
                f.write(str(vars(dvp)))
                f.write("\n")
                f.write(str(hasattr(dvp, 'portKeys')))
                f.write("\n")
                f.write(str(hasattr(dvp, 'config')))
                f.write("\n")
                f.write(str(vars(dvp.config)))
                f.write("\n")
```
I get
```
{'_stub': <pyVmomi.SoapAdapter.SoapStubAdapter instance at 0x7f62a6ea4998>, '_moId': u'distributedvirtualportgroup-14', '_serverGuid': None}
False
True
{'numPorts': 0, 'distributedVirtualSwitch': 'vim.DistributedVirtualSwitch:distributedvirtualswitch-10', 'configVersion': None, 'name': '', 'dynamicType': None, 'dynamicProperty': (vmodl.DynamicProperty) [], 'vendorSpecificConfig': (vim.dvs.KeyedOpaqueBlob) [], 'autoExpand': None, 'key': '', 'scope': (vim.ManagedEntity) [], 'policy': None, 'defaultPortConfig': None, 'vmVnicNetworkResourcePoolKey': None, 'portNameFormat': None, 'type': '', 'uplink': None, 'description': None}
```
So  DistributedVirtualPortgroup returned by vcsim doesn't have portKeys. We will never be in `if hasattr(self.cache.get_network(network_devices[key]['name']), 'portKeys'):`.

I tried after creating a new DVportgroup → same result.

@reversecipher Apply the changed I mentioned [earlier](https://github.com/ansible/ansible/pull/34208#issuecomment-353963384) and for me this PR will be shippable.

reversecipher committed:
Rewrite VDS section

reversecipher committed:
Fix PEP 8 error

tchernomax commented:
shipit

Akasurde commented:
Issue related to DVPortgroup - https://github.com/vmware/govmomi/issues/979

pdellaert reviewed (approved):
LGTM
shipit

Akasurde commented:
rebuild_merge
